diff -rBu --new-file slurm-17.02.9.org/src/slurmd/slurmstepd/step_terminate_monitor.c slurm-17.02.9.new/src/slurmd/slurmstepd/step_terminate_monitor.c
--- slurm-17.02.9.org/src/slurmd/slurmstepd/step_terminate_monitor.c	2018-01-29 09:40:29.000000000 +0100
+++ slurm-17.02.9.new/src/slurmd/slurmstepd/step_terminate_monitor.c	2018-01-29 09:54:21.000000000 +0100
@@ -127,6 +127,7 @@
 		goto done;
 
 	rc = pthread_cond_timedwait(&cond, &lock, &ts);
+    goto done;
 	if (rc == ETIMEDOUT) {
 		char entity[24], time_str[24];
 		time_t now = time(NULL);
